guppylang/std/collections.py

```diff
@@ -85,3 +85,134 @@ def empty_stack() -> Stack[T, MAX_SIZE]:
+            self.buf[i].swap(some((parent_prio, parent_val))).unwrap_nothing()
+            self.buf[parent_i].swap(some((prio, val))).unwrap_nothing()
+            i = parent_i
+        return PriorityQueue(self.buf, self.size + 1)
+
+    @guppy
+    @no_type_check
+    def pop(
+        self: PriorityQueue[T, MAX_SIZE] @ owned,
+    ) -> tuple[int, T, PriorityQueue[T, MAX_SIZE]]:
+        """Removes the next element from the priority queue.
+
+        Panics if the priority queue is empty.
+        """
+        if self.size <= 0:
+            panic("PriorityQueue.pop: priority queue is empty")
+        return_prio, return_val = self.buf[0].take().unwrap()
+        new_size = self.size - 1
+        if new_size == 0:
+            return return_prio, return_val, PriorityQueue(self.buf, new_size)
+        displaced_prio, displaced_val = self.buf[new_size].take().unwrap()
+        i = 0
+        while True:
+            left_i = 2 * i + 1
+            if left_i >= new_size:
+                break
+            right_i = left_i + 1
+            if right_i < new_size:
+                left_elem = self.buf[left_i].take().unwrap()
+                right_elem = self.buf[right_i].take().unwrap()
+                left_prio, left_val = left_elem
+                right_prio, right_val = right_elem
+                if right_prio < left_prio:
+                    child_i, child_prio, child_val = right_i, right_prio, right_val
+                    self.buf[left_i].swap(some((left_prio, left_val))).unwrap_nothing()
+                else:
+                    child_i, child_prio, child_val = left_i, left_prio, left_val
+                    self.buf[right_i].swap(
+                        some((right_prio, right_val))
+                    ).unwrap_nothing()
+            else:
+                left_elem = self.buf[left_i].take().unwrap()
+                child_prio, child_val = left_elem
+                child_i = left_i
+            if displaced_prio <= child_prio:
+                self.buf[child_i].swap(some((child_prio, child_val))).unwrap_nothing()
+                break
+            self.buf[i].swap(some((child_prio, child_val))).unwrap_nothing()
+            i = child_i
+        self.buf[i].swap(some((displaced_prio, displaced_val))).unwrap_nothing()
+        return return_prio, return_val, PriorityQueue(self.buf, new_size)
+
+    @guppy
+    @no_type_check
+    def peek(
+        self: PriorityQueue[TCopyable, MAX_SIZE] @ owned,
+    ) -> tuple[int, TCopyable, PriorityQueue[TCopyable, MAX_SIZE]]:
+        """Returns a copy of the next element in the priority queue without removing it.
+
+        Panics if the priority queue is empty.
+
+        Note that this operation is only allowed if the priority queue elements are
+        copyable.
+        """
+        if self.size <= 0:
+            panic("PriorityQueue.peek: priority queue is empty")
+        elem = self.buf[0].unwrap()
+        return elem[0], elem[1], PriorityQueue(self.buf, self.size)
```

Comment:
We don't have tuple indexing in Guppy (yet), so you have to use unpacking instead (it is generally worth testing all methods you've added to catch cases like this!):
```suggestion
        prio, val = self.buf[0].unwrap()
        return prio, val, PriorityQueue(self.buf, self.size)
```